src/main/java/org/apache/sysml/hops/rewrite/RewriteEMult.java

```diff
@@ -0,0 +1,284 @@
+	 * The order is as follows:
+	 * 		scalars > row vectors > col vectors >
+	 *      non-vector matrices ordered by sparsity (higher nnz first, unknown sparsity last) >
+	 *      other data types.
+	 * Disambiguate by Hop ID.
+	 */
+	private static final Comparator<Hop> compareByDataType = new Comparator<Hop>() {
+		@Override
+		public final int compare(Hop o1, Hop o2) {
+			int c = Integer.compare(orderDataType[o1.getDataType().ordinal()], orderDataType[o2.getDataType().ordinal()]);
+			if (c != 0) return c;
+
+			// o1 and o2 have the same data type
+			switch (o1.getDataType()) {
+			case SCALAR: return Long.compare(o1.getHopID(), o2.getHopID());
+			case MATRIX:
+				// two matrices; check for vectors
+				if (o1.getDim1() == 1) { // row vector
+						if (o2.getDim1() != 1) return 1; // row vectors are greatest of matrices
+						return compareBySparsityThenId(o1, o2); // both row vectors
+				} else if (o2.getDim1() == 1) { // 2 is row vector; 1 is not
+						return -1; // row vectors are the greatest matrices
+				} else if (o1.getDim2() == 1) { // col vector
+						if (o2.getDim2() != 1) return 1; // col vectors greater than non-vectors
+						return compareBySparsityThenId(o1, o2); // both col vectors
+				} else if (o2.getDim2() == 1) { // 2 is col vector; 1 is not
+						return 1; // col vectors greater than non-vectors
+				} else { // both non-vectors
+						return compareBySparsityThenId(o1, o2);
+				}
+			default:
+				return Long.compare(o1.getHopID(), o2.getHopID());
+			}
+		}
+		private int compareBySparsityThenId(Hop o1, Hop o2) {
+			// the hop with more nnz is first; unknown nnz (-1) last
+			int c = Long.compare(o1.getNnz(), o2.getNnz());
+			if (c != 0) return c;
+			return Long.compare(o1.getHopID(), o2.getHopID());
+		}
+		private final int[] orderDataType;
+		{
+			Expression.DataType[] dtValues = Expression.DataType.values();
+			orderDataType = new int[dtValues.length];
+			for (int i = 0, valuesLength = dtValues.length; i < valuesLength; i++) {
+				switch(dtValues[i]) {
+				case SCALAR:
+					orderDataType[i] = 4;
+					break;
+				case MATRIX:
+					orderDataType[i] = 3;
+					break;
+				case FRAME:
+					orderDataType[i] = 2;
+					break;
+				case OBJECT:
+					orderDataType[i] = 1;
+					break;
+				case UNKNOWN:
+					orderDataType[i] = 0;
+					break;
+				}
+			}
+		}
+	};
+
+	/**
+	 * Check if a node has a parent that is not in the set of emults. Recursively check children who are also emults.
+	 * @param emults The set of BinaryOp element-wise multiply hops in the emult chain.
+	 * @param child An interior emult hop in the emult chain dag.
+	 * @return Whether this interior emult or any child emult has a foreign parent.
+	 */
+	private static boolean checkForeignParent(final Set<BinaryOp> emults, final BinaryOp child) {
+		final ArrayList<Hop> parents = child.getParent();
+		if (parents.size() > 1)
+			for (final Hop parent : parents)
+				//noinspection SuspiciousMethodCalls
```

Comment:
`emults` is a set of BinaryOp. `parent` can be any Hop. Intellij warns when calling `contains()` with a different type than the type of the set; in this case Hop instead of BinaryOp. However the call is safe and so I include the statement to suppress the warning.